### DISCUSSION TITLE
Cache misses with hash_dir, base_dir & CMake where builddir != child(source dir)

### DISCUSSION THREAD

**Discussion by nehaljwani** [ccache/ccache#485] [Support]
Directory layout:
```
$ tree .
.
├── build_directories
└── source_checkouts
    ├── 1
    │   ├── CMakeLists.txt
    │   └── main.x.cpp
    └── 2
        ├── CMakeLists.txt
        └── main.x.cpp

$ cat source_checkouts/1/main.x.cpp 
#include <iostream>
int main(int, char**) {

    std::cout << "Hello from example's main.x.cpp" << std::endl;
    return 0;
}
```
source_checkouts is the parent folder where all of my source files are checked out. 1 and 2 are checkouts of the source at the same SHA.
build_directories is the parent folder for all of my builds.

Current directory:
```
$ pwd
/spare/scratch/1572965506/expt
```

ccache config (v3.7.5):
```
base_dir = /spare/scratch/1572965506/expt/source_checkouts
hash_dir = false
```

Scenario 1:
```
export CCACHE_DEBUG=1
cmake -S source_checkouts/1 -B build_directories/b1 -GNinja -DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache
cmake --build build_directories/b1/

cmake -S source_checkouts/2 -B build_directories/b2 -GNinja -DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache
cmake --build build_directories/b2/
```

Diffs in hash computation:
```
$ diff build_directories/b1/CMakeFiles/example_exe.dir/main.x.cpp.o.ccache-input-text build_directories/b2/CMakeFiles/example_exe.dir/main.x.cpp.o.ccache-input-text
24c24
< ../../source_checkouts/1/main.x.cpp
---
> ../../source_checkouts/2/main.x.cpp
44c44
< ../../source_checkouts/1/main.x.cpp
---
> ../../source_checkouts/2/main.x.cpp
53c53
< ../../source_checkouts/1/main.x.cpp
---
> ../../source_checkouts/2/main.x.cpp
29679c29679
< ../../source_checkouts/1/main.x.cpp
---
> ../../source_checkouts/2/main.x.cpp
29683c29683
< ../../source_checkouts/1/main.x.cpp
---
> ../../source_checkouts/2/main.x.cpp
```

Scenario 2:
I move the build directories inside the source directories.
```
export CCACHE_DEBUG=1
cmake -S source_checkouts/1 -B source_checkouts/1/b -GNinja -DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache
cmake --build source_checkouts/1/b

cmake -S source_checkouts/2 -B source_checkouts/2/b -GNinja -DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache
cmake --build source_checkouts/1/b
```

And now:
```
$ diff source_checkouts/{1,2}/b/CMakeFiles/example_exe.dir/main.x.cpp.o.ccache-input-text
$
```

My use case is scenario 1, where I want to have the build directories on separate partition and so they can't be children of the source directories. Is it possible to get cache hits in that scenario? I couldn't find any combinations of ccache parameters to make that happen :(

**Comment by jrosdahl** [comment]
Hi,

The question can be reformulated like this: Why doesn't ccache share cache entries for the following compilations where `A/file.cpp` and `B/file.cpp` are identical?

```
g++ -c A/file.cpp
g++ -c B/file.cpp
```

The answer is that ccache includes the source file path in the cache key since there otherwise could be false cache hits if `file.cpp` includes a `__FILE__` macro, e.g. like this:

```c++
#include <iostream>

int main()
{
  std::cout << __FILE__ << "\n";
}
```

You can opt out of this – see the [`file_macro` under "sloppiness" in the manual](https://github.com/ccache/ccache/blob/d8e2ab9/doc/MANUAL.adoc#user-content-config_sloppiness).

**Comment by djobet** [comment]
Hello, similar setup as the OP here.
Here's my config :

export CCACHE_DIR=/n/local2/djobet/ccache # SSD
export CCACHE_MAXSIZE=20G
export CCACHE_BASEDIR=~/work # NFS
export CCACHE_NOHASHDIR=
export CCACHE_SLOPPINESS=file_macro

My sources consist of
~/work/checkout1
~/work/checkout2 # exact same copy as ~/work/checkout1
~/work/some_common_libs

My builds are in
/spare/local/djobet/build_checkout1
/spare/local/djobet/build_checkout2

~/work is under NFS so I can edit remotely easily
I put the object files under /spare/local so the compilation is quicker

Using CCACHE_SLOPPINESS=file_macro does not make any difference.
In both cases, I get around 30% cache hit (build from scratch with checkout1, then build from checkout2, where optimistically I should get 100% cache hit).
I get 30% cache hit from ~/work/some_common_libs because both checkout1 and checkout2 depends on it.

I think the cache hit is bad because when using BASEDIR, the file paths get rewritten relative to the current working directory. cmake starts by doing a "cd /spare/local/djobet/build_checkout1", so  paths get rewritten as ../../../../\~/work/checkout1/xxx and ../../../../\~/work/checkout2/xxx which cannot be matched.

In my team, it was reported that putting the build dir as a subdirectory of each checkout results in the cache hit increasing to 70%.
In my case this is a no go as I don't want to slow down the build (by storing the object files on NFS).

Is there another trick / settings to make it work ?
If not, do you think this is a setup that could be supported ?

**Comment by llunak** [comment]
The point of the basedir technique is that all builds use the same relative path, so they can be reused, because with the same relative path they are, well, the same. But if the path is once ../../source_checkouts/1/main.x.cpp and then ../../source_checkouts/2/main.x.cpp , how is ccache supposed to make that the same?

How you tried making a symlink from the builddir to the srcdir? Then in both cases the relative path could be srcdir/main.x.cpp .

**Comment by djobet** [comment]
making a symlink : this is something I tried.
Unfortunately, maybe because of NFS, cmake kept on re-generating the project in an infinite loop. (It looked like it was constantly seeing CMakeFiles.txt had been touched).
